In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import train_test_split


---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv(Path('Resources/lending_data.csv'))

# Review the DataFrame
print(df.head())


   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop(columns='loan_status')


In [4]:
# Review the y variable Series
print(y.head())


0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [5]:
# Review the X variable DataFrame
print(X.head())


   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  


### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
print(y.value_counts())


0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_regression_model.fit(X_train, y_train)


LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
y_pred = logistic_regression_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
accuracy_score = accuracy_score(y_test, y_pred)


In [11]:
# Generate a confusion matrix for the model
confusion_matrix = confusion_matrix(y_test, y_pred)


In [12]:
# Print the classification report for the model
classification_report = classification_report(y_test, y_pred)

print("Accuracy Score: ", accuracy_score)
print("Confusion Matrix: ", confusion_matrix)
print("Classification Report: ", classification_report)


Accuracy Score:  0.9918489475856377
Confusion Matrix:  [[18663   102]
 [   56   563]]
Classification Report:                precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The accuracy score of approximately 99.18% indicates that our model is correct in nearly 99 out of 100 cases when making predictions.

The confusion matrix provides us with four essential insights:

The model correctly identified 18,663 instances of healthy loans.
The model correctly identified 563 instances of high-risk loans.
The model incorrectly classified 102 healthy loans as high-risk loans.
The model incorrectly classified 56 high-risk loans as healthy loans.
Regarding predictions for 'healthy loans,' our model performs exceptionally well, exhibiting a high level of accuracy in identifying such cases.

However, when it comes to predictions for 'high-risk loans,' our model's performance is still quite good, but not as stellar as it is with 'healthy loans.' This discrepancy could be attributed to the fact that there are significantly more healthy loans than high-risk ones in our dataset, leading the model to be more proficient in handling the former due to the greater exposure during training.

In conclusion, our model does an impressive job overall. It excels in identifying healthy loans and remains fairly adept at dealing with high-risk ones despite the slight difference in performance.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [15]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler


In [16]:

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)


In [17]:
# Count the distinct values of the resampled labels data
print(y_resampled.value_counts())


0    56271
1    56271
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [18]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
model_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_resampled_pred = model_resampled.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [22]:
from sklearn.metrics import confusion_matrix, classification_report

# Print the balanced_accuracy score of the model 
print("Balanced Accuracy Score: ", balanced_accuracy_score(y_test, y_resampled_pred))


Balanced Accuracy Score:  0.9936781215845847


In [23]:
# Generate a confusion matrix for the model
print("Confusion Matrix: ", confusion_matrix(y_test, y_resampled_pred))


Confusion Matrix:  [[18649   116]
 [    4   615]]


In [24]:
# Print the classification report for the model
print("Classification Report: ", classification_report(y_test, y_resampled_pred))


Classification Report:                precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The pled data reveals an impressive performance of the model in determining the safety (labeled as 0) or high-risk (labeled as 1) status of loans.

Let's start with the safe loans. In this case, the model achieves a perfect precision score of 1.00, indicating that when it classifies a loan as safe, it is almost always accurate. Additionally, the model exhibits a recall of 0.99, which means it can effectively identify nearly all safe loans correctly.

As for the high-risk loans, the model's precision is 0.84, indicating that it correctly identifies about 84% of the high-risk loans it flags. Moreover, the model's recall for high-risk loans is 0.99, showcasing its ability to effectively detect the majority of high-risk loans.

In simpler terms, this logistic regression model excels in identifying both safe and high-risk loans. It demonstrates a high level of accuracy and effectively identifies most of the high-risk loans. Consequently, it proves to be an extremely valuable tool for credit risk analysis.